### ISSUE TITLE
Use cmake, not make directly for compilation, separate source and build?

### ISSUE THREAD

**Issue by Mebus** [teddych/railcontrol#2]
Hallo Teddy,

I noticed something that is not very nice. When building railcontrol you suggest to use just 

```
make
```

which will perform a in-source build of railcontrol. This is not nice. I would recommend to use the following as official build instructions:

```
cmake -B build -S .
camke --build build
```

This doesn't throw *.o files accross the source code and is the recommended way of using cmake. These instructions should be added to the README.md file, so one does not have to open the website to find out how to build railcontrol.

Unfortunately there is an issue with the Version.cpp file with the current CMakeLists.txt file. Can that be fixed?

Mebus

**Comment by teddych** [comment]
Hello Mebus

As on some platforms (i.e. OSX/MacOS) cmake is not easily available I will not recommend cmake as official method to build RailControl. In contrary GNU make is available and widely known on all platforms I know, or at least it is installable easily.

For those they want to build with cmake I have already created a CMakeLists.txt. I will also maintain this file in the future.

Keep in mind that most RailControl-users that are building from source code do not have any programming knowledge. They do it because they want to use the latest version of RailControl that is not already available in a precompiled package.

You are right, it is not very nice to build within the source directory. But unless now I didn't find the time to fix this.

Regards
Teddy

**Comment by Mebus** [comment]
Hello Teddy,

I doubt very heavily that cmake is a problem on OSX. This website:

https://cmake.org/install/

says that binary downloads are available for OSX.

Mebus

**Comment by teddych** [comment]
Hi Mebus

For you as an experienced developer it is not a problem to install an extra tool. But there are a lot of users that are not experienced and are not developers. For them an (unneeded) extra tool is just annoying. cmake is not part of the xcode development environment on OSX. Also on some Linux-distribution - especially small Raspberry-like systems - do not include cmake by default or neither at all. Therefore I will not use cmake as standard way of installing RailControl. Especially because cmake is not a replacement for GNU make, it is just an add on. If one wants to use it, go on and use it, but it is not a requirement and will never be a requirement.

By the way, as an alternative to cmake one can use autotools, scons, premake, ninja, ant and many others. Why not using one of them? All have their advantages and disadvantages. Two disadvantages of all these tools: They have to be installed additionally and one has to manage an extra (usually complicated) configuration.

In short: I can't see the real advantage of cmake or some of the other tools.

Regards
Teddy

**Comment by DB-59** [comment]
Hello friends,

as an user of railcontrol I'd like to give you a very short comment:

I agree to Teddys statement to 100%.

My Background:

I'll try to understand/modify railcontrol in near future. 

At this moment I enjoy it to use railcontrol, helping Teddy for testing and so on, discussing on his forum, learning a lot about the whole MCAN-system (Soft- AND Hardware, commercial AND free).

Teddy has built RailControl on a wide and stable basement - it gives me a very good feeling when using it. Yesterday, today and tomorrow.

With best regards - Detlef, active RailPi-user

(RailPi: RailControl on RaspberryPi with Märklin-Gleisbox)

**Comment by Mebus** [comment]
Ok, do what ever you want, but I don't like to worry about legacy C++11 and ancient Makefiles.

Mebus

**Comment by teddych** [comment]
RailControl has been and still is designed to run on low cost, low powered, cheep and/or old hardware. There are a lot of users that can't or don't want to spend plenty of money for their hobby.

RailControl is especially NOT designed to use all the latest and fanciest techniques. It is not a playground for everything new.

RailControl is a simple and stable software to control model railways.

RailControl depends only on a C++11 compiler, a compatible linker and GNU make. These tools are available on all systems, and on most systems by default. These tools have a stable API that do not change on a daily/weeckly base.

There are other model railway control software (commercial and free) which is based on latest and fanciest tools, with plenty of dependencies to unknown and/or unstable libraries/tools with constantly changing APIs and which requires expensive new hardware. It's up to you to use something other then RailControl. But I don't want to spend my time resolving dependency issues.

Regards
Teddy

**Comment by Mebus** [comment]
GCC supports C++17 since version 8. For instance Debian Buster, which will reach it's end-of-life in July this year, already had GCC version 8. cmake exists since the year 2000. cmake is not playground stuff, but rather an industry standard.

I don't want to make any further comments on the topic. Different people, different opinions.

Mebus
